app/Models/Asset.php

```diff
@@ -981,6 +983,49 @@ public function setNextAuditDateAttribute($value)
         $this->attributes['next_audit_date'] = $value ? Carbon::parse($value)->format('Y-m-d') : null;
     }
 
+    public function getLastCheckoutAttribute($value)
+    {
+        return $this->attributes['last_checkout'] = $value ? Carbon::parse($value)->format('Y-m-d H:i:s') : null;
+    }
+
+    public function setLastCheckoutAttribute($value)
+    {
+        $this->attributes['last_checkout'] = $value ? Carbon::parse($value)->format('Y-m-d H:i:s') : null;
+    }
+
+    public function getLastCheckinAttribute($value)
+    {
+        return $this->attributes['last_checkin'] = $value ? Carbon::parse($value)->format('Y-m-d H:i:s') : null;
+    }
+
+    public function setLastCheckinAttribute($value)
+    {
+        $this->attributes['last_checkin'] = $value ? Carbon::parse($value)->format('Y-m-d H:i:s') : null;
+    }
+
+    public function getLastAuditDateAttribute($value)
+    {
+        return $this->attributes['last_audit_date'] = $value ? Carbon::parse($value)->format('Y-m-d H:i:s') : null;
+    }
+
+    public function setLastAuditDateAttribute($value)
+    {
+        $this->attributes['last_audit_date'] = $value ? Carbon::parse($value)->format('Y-m-d H:i:s') : null;
+    }
+
+    public function getAssetEolDateAttribute($value)
+    {
+        return $this->attributes['asset_eol_date'] = $value ? Carbon::parse($value)->format('Y-m-d') : null;
+    }
+
+    public function setAssetEolDateAttribute($value)
+    {
+        $this->attributes['asset_eol_date'] = $value ? Carbon::parse($value)->format('Y-m-d') : null;
+    }
```

Comment:
If this gets merged we should remove these fields from `$casts` since we're overwriting the functionality. I wouldn't be surprised if keeping them in `$casts` while additionally handling the accessors and mutators ourselves causes a bug down the line because the framework (and/or libraries) assume it is in control of casting the included properties.

Additionally, if this is getting merged, you may want to consider using the new [accessor/mutator syntax](https://laravel.com/docs/10.x/eloquent-mutators#defining-a-mutator) that groups them into one method for each field so it is easier to read.

Reply:
Casting has never worked the way we expected anyway tbh, since it only kicks in when the data is serialized IIRC. It's why we've gone back and forth so many times :(

Reply:
Dates have especially always behaved weirdly in the importer (since that ignores the casting), and I can't help but think that using accessors and mutators is the most consistent path.

Reply:
I _really_ wish we wouldn't do this. A few reasons: 

1. Framework expectations, 90% of the time I'm grabbing a date from a model I'm expecting it to be cast so that I have carbon already available. I'd love to see us work _with_ the framework more often than against it. 
   We should follow framework convention as much as we possibly can - I think that's what will help us iterate more quickly and possibly even see more involvement from the FOSS community. 
2. The getters I feel are actively or have the potential to be actively breaking casting, as Marcus said above.  
3. If in the future we want to use a date-time-picker a little more tightly integrated it likely wouldn't work with this out of the box (think something like filament forms: https://filamentphp.com/docs/3.x/forms/fields/date-time-picker) 
4. On the setter front, with casts, as long as we receive something that's parsable from Carbon we can save it directly to the model instead of having to format it manually. 

> Casting has never worked the way we expected anyway tbh, since it only kicks in when the data is serialized IIRC. It's why we've gone back and forth so many times :(
It's cast as a Carbon object, then the formatting kicks in when an object doesn't make sense (when it's serialized).

Reply:
> Casting has never worked the way we expected anyway tbh, since it only kicks in when the data is serialized IIRC

I was caught off guard by this but it makes sense to me now: you get all the niceties of a carbon object without having to do the casting yourself. So we can compare against other objects or format differently while the model is being passed around the application (etc) and then it falls back to default formatting when leaving the application layer.

Overall, I agree with Spencer's points and think we should lean into the framework as much as possible.

Reply:
> The getters I feel are actively or have the potential to be actively breaking casting, as Marcus said above.

The casting has literally NEVER worked. It's only invoked in JSON and serialization, and we already use a formatting helper for that.

> On the setter front, with casts, as long as we receive something that's parsable from Carbon we can save it directly to the model instead of having to format it manually.

This is literally not what my own tests with this PR have shown. Please, by all means, prove me wrong. Regardless of how I tried to parse the date, if it was blank, it came through as `0000-00-00` or `0000-00-00 00:00:00`, even while checking for bulls, try/catching whether or not Carbon could parse it, etc. 

If I was missing something, I'm glad to take the correction, but this has never worked the way we wanted/expected, and I need stuff that actually works, versus what's "correct".